File: Problems/ImplementQueueUsingStacks.java
/**
 * Implement a first in first out (FIFO) queue using only two stacks. 
 * The implemented queue should support all the functions of a normal queue (push, peek, pop, and empty).
 * 
 * Implement the MyQueue class:
 * 
 * void push(int x) Pushes element x to the back of the queue.
 * int pop() Removes the element from the front of the queue and returns it.
 * int peek() Returns the element at the front of the queue.
 * boolean empty() Returns true if the queue is empty, false otherwise.
 * 
 * Notes:
 * You must use only standard operations of a stack, which means only push to top, peek/pop from top, size, and is empty operations are valid.
 * Depending on your language, the stack may not be supported natively. 
 * You may simulate a stack using a list or deque (double-ended queue) as long as you use only a stack's standard operations.
 */
package Problems;

import java.util.*;

public class ImplementQueueUsingStacks {
    Stack<Integer> stack1 = new Stack<>();
    Stack<Integer> stack2 = new Stack<>();

    public ImplementQueueUsingStacks() {

    }

    public void push(int x) {
        stack1.add(x);
    }

    public int pop() {
        if (stack2.isEmpty()) {
            while (!stack1.isEmpty()) {
                stack2.add(stack1.pop());
            }
            return stack2.pop();
        } else {
            return stack2.pop();
        }
    }

    public int peek() {
        if (stack2.isEmpty()) {
            while (!stack1.isEmpty()) {
                stack2.add(stack1.pop());
            }
            return stack2.peek();
        } else {
            return stack2.peek();
        }
    }

    public boolean empty() {
        return stack1.isEmpty() && stack2.isEmpty();
    }

    public static void main(String[] args) {
        ImplementQueueUsingStacks obj = new ImplementQueueUsingStacks();
        obj.push(1); // queue is: [1]
        obj.push(2); // queue is: [1, 2] (leftmost is front of the queue)
        System.out.println(obj.peek()); // return 1
        System.out.println(obj.pop()); // return 1, queue is [2]
        System.out.println(obj.empty()); // return false
    }
}

File: Problems/ImplementQueueUsingStacks.cpp
#include <iostream>
#include <stack>
using namespace std;

class MyQueue {
   public:
    MyQueue() {}

    void push(int x) {
        pushStack.push(x);
    }

    int pop() {
        int ret;
        if (popStack.empty())
            fillpopStack();
        ret = popStack.top();
        popStack.pop();
        return ret;
    }

    int peek() {
        if (popStack.empty())
            fillpopStack();
        return popStack.top();
    }

    bool empty() {
        return popStack.empty() && pushStack.empty();
    }

   private:
    stack<int> pushStack;
    stack<int> popStack;
    void fillpopStack() {
        while (!pushStack.empty()) {
            popStack.push(pushStack.top());
            pushStack.pop();
        }
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */